File: Assets/src/model/actors/Actor.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Runtime.Serialization;
using UnityEngine;

// Called when the attached actor attacks a target, before damage is dealt.
// damage is after Actor modifiers, but before target modifiers.
public interface IAttackHandler {
  void OnAttack(int damage, Body target);
}

/// Called when Actor deals attack damage (including 0). This includes non-attack sources.
public interface IDealAttackDamageHandler {
  /// damage is after both Actor modifiers and target modifiers
  void HandleDealAttackDamage(int damage, Body target);
}

public interface IActionPerformedHandler {
  void HandleActionPerformed(BaseAction final, BaseAction initial);
}

public interface IStatusAddedHandler {
  void HandleStatusAdded(Status status);
}

public interface IStatusRemovedHandler {
  void HandleStatusRemoved(Status status);
}

[Serializable]
public class Actor : Body, ISteppable {
  public float timeNextAction { get; set; }
  public virtual float turnPriority => 50;

  public static ActionCosts StaticActionCosts = new ActionCosts {
    {ActionType.ATTACK, 1},
    {ActionType.GENERIC, 1},
    {ActionType.MOVE, 1},
    {ActionType.WAIT, 1},
  };

  public override IEnumerable<object> MyModifiers => base.MyModifiers.Concat(statuses.list).Append(this.task);

  public StatusList statuses;
  public override int maxHp => Modifiers.Process(this.MaxHPModifiers(), baseMaxHp);

  /// don't call directly; this doesn't use modifiers
  protected virtual ActionCosts actionCosts => Actor.StaticActionCosts;
  public float baseActionCost => GetActionCost(ActionType.WAIT);
  public virtual ActorTask task {
    get => taskQueue.FirstOrDefault();
    set => SetTasks(value);
  }
  public IEnumerable<ActorTask> tasks => taskQueue.AsEnumerable();

  protected List<ActorTask> taskQueue = new List<ActorTask>();

  [field:NonSerialized] /// Controller only
  public event Action<ActorTask> OnSetTask;
  public Faction faction = Faction.Neutral;
  /// gets called on any ground targeted attack
  [field:NonSerialized] /// Controller only
  public event Action<Vector2Int> OnAttackGround;

  public Actor(Vector2Int pos) : base(pos) {
    statuses = new StatusList(this);
    hp = baseMaxHp = 8;
    // this.timeNextAction = this.timeCreated + baseActionCost;
    this.timeNextAction = this.timeCreated;
  }

  /// create an Attack with the specified damage. This does *not* do damage modifiers.
  internal void Attack(Body target, int damage) {
    if (target.IsDead) {
      throw new CannotPerformActionException("Cannot attack dead target.");
    }
    OnAttack(damage, target);
    target.Attacked(damage, this);
  }

  /// Attack the target, using this Actor's final attack damage.
  internal virtual void Attack(Body target) {
    Attack(target, GetFinalAttackDamage());
  }

  /// get one instance of an attack damage from this Actor
  internal virtual (int, int) BaseAttackDamage() {
    return (0, 0);
  }

  internal virtual int GetFinalAttackDamage() {
    var (min, max) = BaseAttackDamage();
    var baseDamage = MyRandom.Range(min, max + 1);
    var finalDamage = Modifiers.Process(this.AttackDamageModifiers(), baseDamage);
    return finalDamage;
  }

  internal virtual void AttackGround(Vector2Int targetPosition) {
    Body target = floor.bodies[targetPosition];
    Grass grass = floor.grasses[targetPosition];
    OnAttackGround?.Invoke(targetPosition);
    if (target != null) {
      Attack(target);
    } else if (grass != null) {
      // kill the grass when it gets attacked
      grass.Kill(this);
    }
  }

  public override void Kill(Entity source) {
    if (!IsDead) {
      taskQueue.Clear();
      TaskChanged();
      foreach (var handler in Modifiers.Of<IActorKilledHandler>(this).ToList()) {
        handler.OnKilled(this);
      }
      base.Kill(source);
    }
  }

  public void ClearTasks() {
    SetTasks();
  }

  public void SetTasks(params ActorTask[] tasks) {
    taskQueue.Clear();
    taskQueue.AddRange(tasks);
    TaskChanged();
  }

  public void InsertTasks(params ActorTask[] tasks) {
    taskQueue.InsertRange(0, tasks);
    TaskChanged();
  }

  /// Call when this.task is changed
  protected virtual void TaskChanged() {
    OnSetTask?.Invoke(this.task);
  }

  /// uses modifiers
  public float GetActionCost(ActionType t) {
    return Modifiers.Process(this.ActionCostModifiers(), actionCosts.Copy())[t];
  }

  public virtual float GetActionCost(BaseAction action) {
    return GetActionCost(action.Type);
  }

  public virtual float Step() {
    if (task == null) {
      throw new NoActionException();
    }
    task.PreStep();
    /// clear out all done actions from the queue
    while (task.IsDone()) {
      // this mutates action
      GoToNextTask();
      if (task == null) {
        throw new NoActionException();
      } else {
        task.PreStep();
      }
    }

    var isFree = task.isFreeTask;
    // at this point, we know task is not null and it is not done
    var action = task.GetNextAction();
    BaseAction finalAction = Perform(action);
    if (IsDead) {
      throw new ActorDiedException();
    }
    Modifiers.Process(this.StepModifiers(), null);
    task?.PostStep();

    // handle close-ended actions
    while (task != null && (task.WhenToCheckIsDone.HasFlag(TaskStage.After) && !task.forceOnlyCheckBefore) && task.IsDone()) {
      GoToNextTask();
    }

    if (isFree) {
      return 0;
    }
    return GetActionCost(finalAction);
  }

  public BaseAction Perform(BaseAction action) {
    var finalAction = Modifiers.Process(this.BaseActionModifiers(), action);
    finalAction.Perform();
    OnActionPerformed(finalAction, action);
    return finalAction;
  }

  /// Precondition: this.action's enumerator is ended, but the action is still in the queue.
  /// This will remove this.action from the queue and call .Finish() on it, which will
  /// indirectly set this.action to the next one in the queue
  public virtual void GoToNextTask() {
    task.Ended();
    taskQueue.RemoveAt(0);
    TaskChanged();
  }

  private void OnAttack(int damage, Body target) {
    foreach (var handler in this.Of<IAttackHandler>()) {
      handler.OnAttack(damage, target);
    }
  }
  
  private void OnActionPerformed(BaseAction final, BaseAction initial) {
    foreach (var handler in this.Of<IActionPerformedHandler>()) {
      handler.HandleActionPerformed(final, initial);
    }
  }

  /// trigger this Actor dealing attack damage.
  public void OnDealAttackDamage(int damage, Body target) {
    foreach (var handler in this.Of<IDealAttackDamageHandler>()) {
      handler.HandleDealAttackDamage(damage, target);
    }
  }

  public override string ToString() {
    return base.ToString() + $", HP {hp}/{maxHp}, Statuses: {string.Join(", ", statuses)}";
  }

  public bool CanTargetPlayer() {
    return isVisible;
  }
}

public enum Faction { Ally = 1, Neutral = 2, Enemy = 4 }

public class ActionCosts : Dictionary<ActionType, float> {
  public ActionCosts(IDictionary<ActionType, float> dictionary) : base(dictionary) {}
  public ActionCosts() : base() {}
  public ActionCosts Copy() => new ActionCosts(this);
}

public class Attack {
  public readonly Body target;
  public int damage;
  public readonly Actor source;

  public Attack(Actor source, Body target, int damage) {
    this.source = source;
    this.target = target;
    this.damage = damage;
  }

  public Attack(Actor source, Body target) : this(source, target, source.GetFinalAttackDamage()) {
  }
}

// /// First we create an attack "instance", and
// /// fill it up with info.
// /// Then we "execute" the attack
// public class AttackInstance {
//   public AttackInstance(Actor source, Actor target) {
//     this.source = source;
//     this.target = target;
//   }

//   public Actor source { get; }
//   // source's weapon
//   public Item weapon { get; set; }
//   public Actor target { get; }
//   // target's armor
//   public Item armor { get; set; }

//   void Execute() {
//     int damage = source.GetAttackDamage();
//     if (damage < 0) {
//       Debug.LogWarning("Cannot take negative damage!");
//       return;
//     }
//     target.TakeDamage(damage, source);
//   }
// }

File: web/src/model/Actor.ts
import { Body } from './Body';
import { Vector2Int } from '../core/Vector2Int';
import {
  ActionType,
  Faction,
} from '../core/types';
import {
  collectModifiers,
  processModifiers,
  ACTION_COST_MOD,
  BASE_ACTION_MOD,
  ATTACK_DAMAGE_MOD,
  STEP_MOD,
  MAX_HP_MOD,
  type IActionCostModifier,
  type IBaseActionModifier,
  type IAttackDamageModifier,
  type IStepModifier,
  type IMaxHPModifier,
} from '../core/Modifiers';
import { MyRandom } from '../core/MyRandom';
import { EventEmitter } from '../core/EventEmitter';
import { BaseAction, ActionCosts } from './BaseAction';
import { GameModelRef } from './GameModelRef';
import { ActorTask, TaskStage } from './ActorTask';
import { StatusList } from './StatusList';
import type { ISteppable } from './Floor';

// ─── Handler symbols ───
export const ATTACK_HANDLER = Symbol.for('IAttackHandler');
export const DEAL_ATTACK_DAMAGE_HANDLER = Symbol.for('IDealAttackDamageHandler');
export const ACTION_PERFORMED_HANDLER = Symbol.for('IActionPerformedHandler');
export const ACTOR_KILLED_HANDLER = Symbol.for('IActorKilledHandler');
export const STATUS_ADDED_HANDLER = Symbol.for('IStatusAddedHandler');
export const STATUS_REMOVED_HANDLER = Symbol.for('IStatusRemovedHandler');

export interface IAttackHandler {
  onAttack(damage: number, target: Body): void;
}
export interface IDealAttackDamageHandler {
  handleDealAttackDamage(damage: number, target: Body): void;
}
export interface IActionPerformedHandler {
  handleActionPerformed(finalAction: BaseAction, initialAction: BaseAction): void;
}

export class NoActionException extends Error {
  constructor() {
    super('No action available');
    this.name = 'NoActionException';
  }
}

export class ActorDiedException extends Error {
  constructor() {
    super('Actor died during step');
    this.name = 'ActorDiedException';
  }
}

/**
 * Body with tasks, statuses, combat, and turn-stepping.
 * Port of C# Actor.cs.
 */
export class Actor extends Body implements ISteppable {
  timeNextAction: number;
  get turnPriority(): number {
    return 50;
  }

  readonly statuses: StatusList;
  faction: Faction = Faction.Neutral;

  protected taskQueue: ActorTask[] = [];

  readonly onSetTask = new EventEmitter<[ActorTask | null]>();
  readonly onAttackGround = new EventEmitter<[Vector2Int]>();
  readonly afterActionPerformed = new EventEmitter<[BaseAction, BaseAction]>();

  get myModifiers(): Iterable<object | null | undefined> {
    return [...super.myModifiers, ...this.statuses.list, this.task ?? null];
  }

  get maxHp(): number {
    const mods = collectModifiers<IMaxHPModifier>(this, MAX_HP_MOD);
    return processModifiers(mods, this._baseMaxHp);
  }

  /** Stationary actors never move — suppresses idle bob animation. Override to true in subclasses. */
  get isStationary(): boolean {
    return false;
  }

  protected get actionCosts(): ActionCosts {
    return ActionCosts.default();
  }

  get baseActionCost(): number {
    return this.getActionCost(ActionType.WAIT);
  }

  get task(): ActorTask | null {
    return this.taskQueue[0] ?? null;
  }

  set task(value: ActorTask | null) {
    if (value) {
      this.setTasks(value);
    } else {
      this.clearTasks();
    }
  }

  get tasks(): readonly ActorTask[] {
    return this.taskQueue;
  }

  constructor(pos: Vector2Int) {
    super(pos);
    this.statuses = new StatusList(this);
    this._hp = this._baseMaxHp = 8;
    this.timeNextAction = this.timeCreated;
  }

  /** Attack the target with the specified raw damage (no modifiers). */
  attackWithDamage(target: Body, damage: number): void {
    if (target.isDead) {
      throw new Error('Cannot attack dead target.');
    }
    GameModelRef.mainOrNull?.emitAnimation({ type: 'attack', entityGuid: this.guid, from: this.pos, to: target.pos, targetGuid: target.guid, amount: damage });
    this.onAttackEvent(damage, target);
    target.attacked(damage, this);
  }

  /** Attack the target, using this Actor's final attack damage. */
  attack(target: Body): void {
    this.attackWithDamage(target, this.getFinalAttackDamage());
  }

  /** Get base attack damage range [min, max]. Override in subclasses. */
  baseAttackDamage(): [number, number] {
    return [0, 0];
  }

  getFinalAttackDamage(): number {
    const [min, max] = this.baseAttackDamage();
    const baseDamage = MyRandom.Range(min, max + 1);
    const mods = collectModifiers<IAttackDamageModifier>(this, ATTACK_DAMAGE_MOD);
    return processModifiers(mods, baseDamage);
  }

  attackGround(targetPosition: Vector2Int): void {
    const target = this.floor?.bodies.get(targetPosition);
    const grass = this.floor?.grasses.get(targetPosition);
    // Suppress attackGround event when hitting a body — the 'attack' event from this.attack() handles the bump.
    if (!target) {
      GameModelRef.mainOrNull?.emitAnimation({ type: 'attackGround', entityGuid: this.guid, from: this.pos, to: targetPosition });
    }
    this.onAttackGround.emit(targetPosition);
    if (target) {
      this.attack(target as Body);
    } else if (grass) {
      grass.kill(this);
    }
  }

  kill(source: Entity): void {
    if (!this.isDead) {
      this.taskQueue.length = 0;
      this.taskChanged();
      const mods = collectModifiers<{ onKilled(a: Actor): void }>(this, ACTOR_KILLED_HANDLER);
      for (const handler of [...mods]) {
        handler.onKilled(this);
      }
      super.kill(source);
    }
  }

  clearTasks(): void {
    this.setTasks();
  }

  setTasks(...tasks: ActorTask[]): void {
    this.taskQueue.length = 0;
    this.taskQueue.push(...tasks);
    this.taskChanged();
  }

  insertTasks(...tasks: ActorTask[]): void {
    this.taskQueue.unshift(...tasks);
    this.taskChanged();
  }

  protected taskChanged(): void {
    this.onSetTask.emit(this.task);
  }

  /** Get action cost with modifiers applied. */
  getActionCost(t: ActionType): number {
    const mods = collectModifiers<IActionCostModifier>(this, ACTION_COST_MOD);
    const costs = processModifiers(mods, this.actionCosts.copy());
    return costs.get(t) ?? 1;
  }

  getActionCostForAction(action: BaseAction): number {
    return this.getActionCost(action.type);
  }

  step(): number {
    if (!this.task) throw new NoActionException();

    this.task.preStep();

    // Clear out all done tasks
    while (this.task?.isDone()) {
      this.goToNextTask();
      if (!this.task) throw new NoActionException();
      this.task.preStep();
    }

    const isFree = this.task.isFreeTask;
    const action = this.task.getNextAction();
    const finalAction = this.perform(action);

    if (this.isDead) throw new ActorDiedException();

    // Process step modifiers (status effects etc.)
    const stepMods = collectModifiers<IStepModifier>(this, STEP_MOD);
    processModifiers(stepMods, {} as object);

    this.task?.postStep(action, finalAction);

    // Handle close-ended tasks
    while (
      this.task &&
      (this.task.whenToCheckIsDone & TaskStage.After) !== 0 &&
      !this.task.forceOnlyCheckBefore &&
      this.task.isDone()
    ) {
      this.goToNextTask();
    }

    if (isFree) return 0;
    return this.getActionCostForAction(finalAction);
  }

  perform(action: BaseAction): BaseAction {
    const mods = collectModifiers<IBaseActionModifier>(this, BASE_ACTION_MOD);
    const finalAction = processModifiers(mods, action);
    finalAction.perform();
    this.onActionPerformed(finalAction, action);
    return finalAction;
  }

  goToNextTask(): void {
    this.task?.ended();
    this.taskQueue.shift();
    this.taskChanged();
  }

  /** Called when this actor deals attack damage to a target. */
  onDealAttackDamage(damage: number, target: Body): void {
    const handlers = collectModifiers<IDealAttackDamageHandler>(this, DEAL_ATTACK_DAMAGE_HANDLER);
    for (const handler of handlers) {
      handler.handleDealAttackDamage(damage, target);
    }
  }

  canTargetPlayer(): boolean {
    return this.isVisible;
  }

  toString(): string {
    return `${super.toString()}, HP ${this._hp}/${this.maxHp}`;
  }

  // ─── Private event dispatchers ───

  private onAttackEvent(damage: number, target: Body): void {
    const handlers = collectModifiers<IAttackHandler>(this, ATTACK_HANDLER);
    for (const handler of handlers) {
      handler.onAttack(damage, target);
    }
  }

  private onActionPerformed(finalAction: BaseAction, initialAction: BaseAction): void {
    const handlers = collectModifiers<IActionPerformedHandler>(this, ACTION_PERFORMED_HANDLER);
    for (const handler of handlers) {
      handler.handleActionPerformed(finalAction, initialAction);
    }
    this.afterActionPerformed.emit(finalAction, initialAction);
  }

  // ISteppable
  catchUpStep(_lastTime: number, _currentTime: number): void {}
}

// Re-export for convenience
import { Entity } from './Entity';
